keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2508 @@
+	ResourceClassName string
+
+	// Parameters references a separate object with arbitrary parameters
+	// that will be used by the
+	// driver when allocating a resource for the claim.
+	//
+	// The object must be in the same namespace as the ResourceClaim.
+	Parameters ResourceClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. Waiting for a Pod is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod (AllocationModeWaitForFirstConsumer). Other modes
+// might get added in the future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started. It's necessary to enable usage
+	// of the claim by kubelet in case that the ResourceClass got
+	// removed in the meantime. It also helps the resource driver
+	// to determine whether it needs to handle a claim that got
+	// marked for deletion.
+	DriverName string
+
+	// Scheduling contains information that is only relevant while the
+	// scheduler and the resource driver are in the process of selecting a
+	// node for a Pod and the allocation mode is AllocationModeWaitForFirstConsumer. The
+	// resource driver should unset this when it has successfully allocated
+	// the resource.
+	Scheduling SchedulingStatus
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// Deallocate may be set to true to request deallocation of a resource as soon
+	// as it is unused. The scheduler uses this when it finds that deallocating
+	// the resource and reallocating it elsewhere might unblock a pod.
+	//
+	// The resource driver checks this fields and resets it to false
+	// together with clearing the Allocation field. It also sets it
+	// to false when the resource is not allocated.
+	Deallocate bool
```

Comment:
This field is going to be something to watch out for during implementation. But good enough as a high level idea.